hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/AuthenticationBaseMessageTask.java

```diff
@@ -314,21 +286,67 @@ public Permission getRequiredPermission() {
         return null;
     }
 
-    private class OperationInfo {
-        private final Member member;
-        private final Future future;
+    private final class MultiTargetCallback {
 
-        public OperationInfo(Future future, Member member) {
-            this.future = future;
-            this.member = member;
+        private final Collection<Member> targets;
+        private final Map<Member, Object> results;
+
+        private MultiTargetCallback(Collection<Member> targets) {
+            this.targets = synchronizedSet(new HashSet<Member>(targets));
+            this.results = new ConcurrentHashMap<Member, Object>(targets.size());
+        }
+
+        public void notify(Member member, Object result) {
+            if (targets.remove(member)) {
+                results.put(member, result);
+            } else {
+                if (results.containsKey(member)) {
+                    logger.warning("Duplicate response from -> " + member);
+                }
+                logger.warning("Unknown target! -> " + member);
+            }
+            if (!targets.isEmpty()) {
```

Comment:
Then we have this problem in other places too similar implementation in `AbstractMultiTargetMessageTask`. I will try to merge this MultiTargetCallback and SingleTargetCallback with the ones in AbstractMultiTargetMessageTask and remove the race so that it is fixed there too.